ocaml/src/alba_client_download.ml

```diff
@@ -281,13 +292,53 @@ let download_chunk
   let fragments = Hashtbl.create n in
 
   let module CountDownLatch = Lwt_extra2.CountDownLatch in
-  let successes = CountDownLatch.create ~count:k in
-  let failures = CountDownLatch.create ~count:(m+1) in
+  let chunk_locations_i =
+    List.mapi
+      (fun i cl -> (i,cl))
+      chunk_locations
+  in
+  begin
+    Lwt_log.debug_f "download_strategy:%s"
+                  (show_download_strategy download_strategy) >>= fun () ->
+    match download_strategy with
+    | AllFragments -> Lwt.return (chunk_locations_i, k, m+1)
+    | LeastAmount  ->
+       if k = 1
+       then
+         begin
+           Lwt_log.debug_f
+             "replication: opportunity to use read_preference:%s"
+             ([%show: string list option] read_preference)
+           >>= fun () ->
+           (match read_preference with
+            | None-> Lwt.return_none
+            | Some prefered_nodes ->
+               Alba_client_common.find_prefered_osd
+                 prefered_nodes
+                 osd_access
+                 chunk_locations_i
+           )
+           >>= fun prefered_osd_o ->
+
+           let target =
+             match prefered_osd_o with
+             | None -> List.take k chunk_locations_i (* TODO:maybe randomize? *)
+             | Some c_i -> [c_i]
```

Comment:
The since the position of each element is in the `chunk_locations_i` list, the type could be changed from `osd_id option` to `osd_id, but it means a signature change for everything downstream (... down to `Alba_client_download.download_packed_fragment`)